DataFormats/TauReco/src/PFTau.cc

```diff
@@ -121,8 +121,111 @@ namespace {
     oCache.reset();
     oCache.set( std::make_unique<T>(std::move(iFrom)));
   }
+
+  std::unique_ptr<reco::PFCandidatePtr> convertToPFPtr(const reco::CandidatePtr& ptr) {
+    if (ptr.isNonnull()) {
+      const reco::PFCandidate* pf_cand = dynamic_cast<const reco::PFCandidate*>(&*ptr);
+      if (pf_cand != nullptr) {
+        return std::unique_ptr<reco::PFCandidatePtr>(new reco::PFCandidatePtr(ptr));
+      } else throw cms::Exception("Type Mismatch") << "This PFTau was not made from PFCandidates, but it is being tried to access a PFCandidate.\n";
+    }
+    return std::unique_ptr<reco::PFCandidatePtr>(new reco::PFCandidatePtr());
+  }
+
+  std::unique_ptr<std::vector<reco::PFCandidatePtr> > convertToPFPtrs(const std::vector<reco::CandidatePtr>& cands) {
+    std::unique_ptr<std::vector<reco::PFCandidatePtr> > newSignalPFCands{new std::vector<reco::PFCandidatePtr>{}};
+    bool isPF = false;
+    for (auto& cand : cands) {
+      // Check for first Candidate if it is a PFCandidate; if yes, skip for the rest
+      if (!isPF) {
+        const reco::PFCandidate* pf_cand = dynamic_cast<const reco::PFCandidate*>(&*cand);
+        if (pf_cand != nullptr) {
+          isPF = true;
+          newSignalPFCands->reserve(cands.size());
+        } else throw cms::Exception("Type Mismatch") << "This PFTau was not made from PFCandidates, but it is being tried to access PFCandidates.\n";
+      }
+      const auto& newPtr = edm::Ptr<reco::PFCandidate>(cand);
+      newSignalPFCands->push_back(newPtr);
+    }
+    return newSignalPFCands;
+  }
+}
+
+const PFCandidatePtr PFTau::leadPFChargedHadrCand() const {
+  if (!leadPFChargedHadrCand_.isSet())
+    leadPFChargedHadrCand_.set(std::move(convertToPFPtr(leadChargedHadrCand_)));
```

Comment:
```suggestion
    leadPFChargedHadrCand_.set(convertToPFPtr(leadChargedHadrCand_));
```
The use of `std::move` is not necessary and is interfering with an important compiler optimization